python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,576 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import warnings
+
+from cudf import Series
+from sklearn.feature_extraction.text import ENGLISH_STOP_WORDS
+from string import punctuation
+from functools import partial
+import nvtext
+import cupy as cp
+import numbers
+import cudf
+from cuml.common.type_utils import CUPY_SPARSE_DTYPES
+from cuml.prims.label import make_monotonic
+
+
+def _preprocess(doc, lower=False, remove_punctuation=False, delimiter=' '):
+    """Chain together an optional series of text preprocessing steps to
+    apply to a document.
+    Parameters
+    ----------
+    doc: nvstrings
+        The string to preprocess
+    lower: bool
+        Whether to use str.lower to lowercase all of the text
+    remove_punctuation: bool
+        Whether to remove all punctuation from the text before tokenizing.
+        Punctuation characters are taken from string.punctuation
+    Returns
+    -------
+    doc: nvstrings
+        preprocessed string
+    """
+    if lower:
+        doc = doc.lower()
+    if remove_punctuation:
+        punctuation_list = Series(list(punctuation))._column.nvstrings
+        doc = doc.replace_multi(punctuation_list, delimiter, regex=False)
+    return doc
+
+
+class _VectorizerMixin:
+    """Provides common code for text vectorizers (tokenization logic)."""
+    def _remove_stop_words(self, doc):
+        """Remove stop words only if needed."""
+        if self.analyzer == 'word' and self.stop_words is not None:
+            stop_words = Series(self._get_stop_words())._column.nvstrings
+            doc = nvtext.replace_tokens(doc, stop_words, self.delimiter)
```

Comment:
I'm assuming all of the nvtext things will be replaced before merge?

Reply:
Yes, this is the plan, only replace_tokens is missing now.

Reply:
I might be mistaken, but in my test proposed `CountVectorizer` runs about 70x slower than that of sklearn's:

```
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer
X, y = fetch_20newsgroups(return_X_y=True)
cv_sklearn = CountVectorizer(token_pattern=r'\b.+\b')
X_cv_sklearn = cv_sklearn.fit_transform(X)
print(X_cv_sklearn.shape)
(11314, 277786)
```
```
%%timeit
X_cv_sklearn = cv_sklearn.fit_transform(X)
961 ms ± 35.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
```
from cudf import Series
from cuml.feature_extraction.text import CountVectorizer
from time import time
X_cudf = Series(X)
cv_cuml = CountVectorizer()
start = time()
X_cv_cuml = cv_cuml.fit_transform(X_cudf)
print(time() - start)
71.41981506347656
```
```
print(X_cv_cuml.shape)
(11314, 235683)
```